### PR TITLE
add rpn recall function and class-aware rpn

### PR DESCRIPTION
**Pull request by Noplz**

### CODE REVIEW

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py [hunk 1]
```diff
@@ -147,12 +149,21 @@ def get_proposals(self, body_feats, im_info, mode='train'):
...
 
-        rpn_cls_score_prob = fluid.layers.sigmoid(
-            rpn_cls_score, name='rpn_cls_score_prob')
-
+        if self.num_classes == 1:
+            rpn_cls_prob = fluid.layers.sigmoid(
+                rpn_cls_score, name='rpn_cls_prob')
+        else:
+            rpn_cls_score = fluid.layers.transpose(rpn_cls_score, perm=[0, 2, 3, 1])
+            rpn_cls_score = fluid.layers.reshape(rpn_cls_score, shape=(0, 0, 0, -1, self.num_classes))
```
Comment (jerrywgz) [inline PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py]: 跑下pre-commit run 这里python一行太长了，pre-commit会做自动修复

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py [hunk 2]
```diff
@@ -196,20 +207,37 @@ def get_loss(self, im_info, gt_box, is_crowd):
...
+                    anchor_var=anchor_var,
+                    gt_boxes=gt_box,
+                    is_crowd=is_crowd,
+                    im_info=im_info)
+            score_tgt = fluid.layers.cast(x=score_tgt, dtype='float32')
+            score_tgt.stop_gradient = True
+            rpn_cls_loss = fluid.layers.sigmoid_cross_entropy_with_logits(
+                x=score_pred, label=score_tgt)
+        else:
+            score_pred, loc_pred, score_tgt, loc_tgt, bbox_weight = \
```
Comment (jerrywgz) [inline PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py:232]: rpn_target_assign可以复用

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py [hunk 3]
```diff
@@ -254,6 +282,7 @@ class FPNRPNHead(RPNHead):
     ]
 
     def __init__(self,
+                 num_classes=1,
```
Comment (jerrywgz) [inline PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py]: 加在最后一个input吧

File: PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py [hunk 4]
```diff
@@ -41,6 +41,28 @@ def clip_bbox(bbox):
     ymax = max(min(bbox[3], 1.), 0.)
     return xmin, ymin, xmax, ymax
 
+def proposal_eval(results, anno_file, outfile, max_dets=(100, 300, 1000)):
+    assert 'proposal' in results[0]
+    assert outfile.endswith('.json')
+
+    xywh_results = proposal2out(results)
```
Comment (jerrywgz) [inline PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py:54]: 这里是否会出现results为空？如果有的话给我报错信息吧

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py (hunk 2 above)
Comment (Noplz) [reply PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py:232]: softmax 需要额外给gt_labels  第231行

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py (hunk 3 above)
Comment (Noplz) [reply PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py]: done

File: PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py (hunk 1 above)
Comment (Noplz) [reply PaddleCV/PaddleDetection/ppdet/modeling/anchor_heads/rpn_head.py]: done

File: PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py (hunk 4 above)
Comment (Noplz) [reply PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py:54]: done

File: PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py [hunk 5]
```diff
@@ -42,6 +47,34 @@ def clip_bbox(bbox):
...
 
 
+def proposal_eval(results, anno_file, outfile, max_dets=(100, 300, 1000)):
+    assert 'proposal' in results[0]
+    assert outfile.endswith('.json')
+
+    xywh_results = proposal2out(results)
+    assert len(
+        segms_results) > 0, "The number of valid proposal detected is zero.\n \
+        Please use reasonable model and check input data."
```
Comment (jerrywgz) [inline PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py:57]: segms_results -> xywh_results

Comment (Noplz) [reply PaddleCV/PaddleDetection/ppdet/utils/coco_eval.py:57]: done

**Review by jerrywgz — APPROVED** [review]
LGTM
